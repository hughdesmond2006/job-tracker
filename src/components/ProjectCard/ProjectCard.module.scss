$lightText: #adadad;
$green: #69A780;
$lightGreen: #7ac39a;

.detailsArea {
  position: relative;
  padding: 8px;
  border: 1px solid #c9c9c9;
  border-radius: 4px;
  margin-bottom: 8px;
  background-color: rgba(255, 255, 255, 0.48);
  transition: background-color 300ms ease-out;
}

.detailsArea:hover {
  background: #fbfbfb;
  transition: none;
}

.completeButton {
  position: absolute;
  border-radius: 4px;
  background-color: $green;
  color: white;
  cursor: pointer;
  font-weight: 200;
  padding: 4px;
  bottom: 12px;
  left: 50%;
  width: 160px;
  margin-left: -80px;
}

.completeButton:hover {
  background-color: $lightGreen;
}

.id{
  float: left;
}

.date{
  float: right;
}

.id, .date{
  font-weight: 200;
  color: $lightText;
  font-size: .8rem;
}

.job{
  margin-top: 26px;
  text-align: center;
}

.address{
  margin-top: 6px;
  text-align: center;
  margin-bottom: 50px;
}

.completed, .completed:hover{
  background: #87A3AE;
  color: white;
  border: none;
  cursor: default;
}

.title{
  display: inline-block;
  padding: 0;
  margin: 0;
  color: #396677;
}

.titleCompleted, .titleCompleted:hover{
  color: #eedcc7;
}





