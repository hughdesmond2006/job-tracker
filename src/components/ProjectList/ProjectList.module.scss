$gold: #E7BD90;
$lightGold: #eec497;

.list{
    list-style: none;
    margin: 0 auto;
    width: 400px;
    padding: 8px 0 0px;
    @media not screen and (min-width: 400px){
        width: 100%;
    }
}

.reset {
    margin: 10px;
    border-radius: 3px;
    background-color: $gold;
    color: white;
    cursor: pointer;
    font-weight: 200;
    font-size: .9rem;
    padding: 4px;
    left: 0;
    width: 60px;
    position: fixed;
    z-index: 1;
}

.reset:hover {
    background: $lightGold;
}