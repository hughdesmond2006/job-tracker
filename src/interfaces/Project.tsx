export default interface Project{
  id: number,
  name: string,
  address: string[],
  startDate: string,
  completed: boolean
}