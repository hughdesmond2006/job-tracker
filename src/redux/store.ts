import { applyMiddleware, createStore } from "redux";
//import logger from "redux-logger";
import thunk from "redux-thunk";

import projectsReducer from "./reducers/projectReducer";

export default createStore(projectsReducer, applyMiddleware(/*logger,*/ thunk));