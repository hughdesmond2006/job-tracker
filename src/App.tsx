import React from "react";
import "./App.css";
import ProjectList from "./components/ProjectList/ProjectList";
import { Provider } from "react-redux";
import store from "./redux/store";

const App: React.FC = () => {
  return (
    <Provider store={store}>
      <div className="App">
        <ProjectList />
      </div>
    </Provider>
  );
};

export default App;
