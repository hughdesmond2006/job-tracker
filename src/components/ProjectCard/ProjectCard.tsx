import React, { PureComponent } from "react";
import Project from "../../interfaces/Project";
import styles from "./ProjectCard.module.scss";
import store from "../../redux/store";
import { completeProjectBegin } from "../../redux/actions/projectActions";

class ProjectCard extends PureComponent<Project> {
  markCompleted = (id: number): void => {
    store.dispatch(completeProjectBegin(id));
  };

  render() {
    const { id, name, address, startDate, completed } = this.props;

    const completedClass = completed ? " " + styles.completed : "";
    const titleCompletedClass = completed ? " " + styles.titleCompleted : "";

    return (
      <div className={styles.detailsArea + completedClass}>
        <div className={styles.id + completedClass}>#{id}</div>
        <div className={styles.date + completedClass}>
          {new Date(startDate).toLocaleDateString("en-GB")}
        </div>
        <div className={styles.job}>
          <div className={styles.title + titleCompletedClass}>Project</div> {name}
        </div>
        <div className={styles.address}>
          <div className={styles.title + titleCompletedClass}>Address</div>{" "}
          {address.toString()}
        </div>
        {completed ? (
          <div className={styles.completeButton + completedClass}>
            Completed!
          </div>
        ) : (
          <div
            className={styles.completeButton}
            onClick={this.markCompleted.bind(this, id)}
          >
            Mark as Complete!
          </div>
        )}
      </div>
    );
  }
}

export default ProjectCard;
