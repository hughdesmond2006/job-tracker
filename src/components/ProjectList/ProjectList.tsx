import React, { Component } from "react";
import { connect } from "react-redux";
import {fetchProjectsBegin, resetBegin} from "../../redux/actions/projectActions";
import store from "../../redux/store";

import Project from "../../interfaces/Project";
import ProjectCard from "../ProjectCard/ProjectCard";
import styles from "./ProjectList.module.scss";
import Spinner from "../Spinner/Spinner";
import AppState from "../../interfaces/AppState";

export class ProjectList extends Component<AppState> {
  componentDidMount() {
    store.dispatch(fetchProjectsBegin());
  }

  reset = (): void => {
    store.dispatch(resetBegin());
  };

  sortNewestFirst = (project1: Project, project2: Project) => {
    return new Date(project1.startDate) > new Date(project2.startDate) ? -1 : 1;
  };

  sortIncompleteFirst = (project1: Project, project2: Project) => {
    return !project1.completed && project2.completed ? -1 : 1;
  };

  render() {
    const { projects, loading, error } = this.props;

    return (
      <React.Fragment>
        <div className={styles.reset} onClick={this.reset}>
          Reset
        </div>
        {loading ? (
          <Spinner />
        ) : (
          <ul className={styles.list}>
            {projects
              .sort(this.sortNewestFirst)
              .sort(this.sortIncompleteFirst)
              .map((project: Project) => {
                return (
                  <li key={project.id}>
                    <ProjectCard
                      id={project.id}
                      name={project.name}
                      address={project.address}
                      startDate={project.startDate}
                      completed={project.completed}
                    />
                  </li>
                );
              })}
          </ul>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  projects: state.projects,
  loading: state.loading,
  error: state.error
});

export default connect(mapStateToProps)(ProjectList);
