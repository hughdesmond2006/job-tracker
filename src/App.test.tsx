import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import Enzyme, { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import configureStore, { MockStore } from "redux-mock-store";

import ProjectCard from "./components/ProjectCard/ProjectCard";
import ProjectList from "./components/ProjectList/ProjectList";
import store from "./redux/store";
import { Provider } from "react-redux";
import Project from "./interfaces/Project";
import {
  completeProjectBegin,
  fetchProjectsBegin,
  resetBegin
} from "./redux/actions/projectActions";
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() });

describe("<App />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});

const testCard1 = (
  <ProjectCard
    id={12}
    name={"A nice Job"}
    address={["Test Address", "Test Street"]}
    startDate={"2019-04-12T00:00:00.000Z"}
    completed={false}
  />
);

const testCard2 = (
  <ProjectCard
    id={4}
    name={"3 Bedroom Refurb"}
    address={["159 Amiens Street", "Dublin 13"]}
    startDate={"2019-04-23T00:00:00.000"}
    completed={true}
  />
);

describe("<ProjectCard />", () => {
  //smoke
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(testCard1, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  // make sure the correct details are rendered and
  // in the correct format
  it("renders an incomplete project card", () => {
    const projectCard = shallow(testCard1);
    expect(projectCard).toMatchSnapshot();
  });

  // tests to see a complete project is rendered appropriately
  it("renders a complete project card", () => {
    const projectCard = shallow(testCard2);
    expect(projectCard).toMatchSnapshot();
  });
});

const testProjectList = [
  {
    address: ["217 East Wall Road", "Dublin 19"],
    completed: false,
    id: 9,
    name: "Living Room Update",
    startDate: "2019-04-24T00:00:00.000Z"
  },
  {
    address: ["198 Fishamble Street", "Dublin 10"],
    completed: true,
    id: 6,
    name: "Window Replacements",
    startDate: "2019-04-13T00:00:00.000Z"
  },
  {
    address: ["99 Mountjoy Square", "Dublin 1"],
    completed: false,
    id: 7,
    name: "3 Bedroom Refurb",
    startDate: "2019-04-13T00:00:00.000Z"
  }
];

const initialState: {
  projects: Project[];
  loading: boolean;
  error: Error;
} = {
  projects: [...testProjectList],
  loading: false,
  error: null
};

describe("<ProjectList />", () => {
  //smoke
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider store={store}>
        <ProjectList />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  // this tests the loading animation is working
  it("renders the loading spinner", () => {
    const projectList = mount(
      <Provider store={store}>
        <ProjectList />
      </Provider>
    );

    const spinnerClass = projectList.find("div.spinner");

    expect(spinnerClass).toMatchSnapshot();
  });

  // this ensures the list is rendered properly and that the
  // incomplete projects appear first
  it("renders populated list", () => {
    const mockStore = configureStore();
    let store: MockStore;
    let wrapper: ShallowWrapper;

    // setup mount with mock redux store before each test
    store = mockStore(initialState);
    wrapper = shallow(
      <Provider store={store}>
        <ProjectList />
      </Provider>
    );

    wrapper.setState({...initialState});

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("Redux", () => {
  const mockStore = configureStore();
  let store: MockStore;
  let wrapper: ReactWrapper;

  // setup mount with mock redux store before each test
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <ProjectList {...initialState} />
      </Provider>
    );
  });

  it("has props matching with initialState", () => {
    expect(wrapper.find(ProjectList).prop("projects")).toEqual(
      initialState.projects
    );
    expect(wrapper.find(ProjectList).prop("loading")).toEqual(
      initialState.loading
    );
    expect(wrapper.find(ProjectList).prop("error")).toEqual(initialState.error);
  });

  it("can dispatch actions", () => {
    let action;
    store.dispatch(fetchProjectsBegin());
    store.dispatch(completeProjectBegin(1));
    store.dispatch(resetBegin());

    action = store.getActions();

    expect(action[0].type).toBe("FETCH_PROJECTS_BEGIN");
    expect(action[1].type).toBe("COMPLETE_PROJECT_BEGIN");
    expect(action[2].type).toBe("RESET_BEGIN");
  });
});
