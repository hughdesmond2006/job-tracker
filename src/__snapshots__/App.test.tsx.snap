// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard /> renders a complete project card 1`] = `
<div
  className="detailsArea completed"
>
  <div
    className="id completed"
  >
    #
    4
  </div>
  <div
    className="date completed"
  >
    4/23/2019
  </div>
  <div
    className="job"
  >
    <div
      className="title titleCompleted"
    >
      Project
    </div>
     
    3 Bedroom Refurb
  </div>
  <div
    className="address"
  >
    <div
      className="title titleCompleted"
    >
      Address
    </div>
     
    159 Amiens Street,Dublin 13
  </div>
  <div
    className="completeButton completed"
  >
    Completed!
  </div>
</div>
`;

exports[`<ProjectCard /> renders an incomplete project card 1`] = `
<div
  className="detailsArea"
>
  <div
    className="id"
  >
    #
    12
  </div>
  <div
    className="date"
  >
    4/12/2019
  </div>
  <div
    className="job"
  >
    <div
      className="title"
    >
      Project
    </div>
     
    A nice Job
  </div>
  <div
    className="address"
  >
    <div
      className="title"
    >
      Address
    </div>
     
    Test Address,Test Street
  </div>
  <div
    className="completeButton"
    onClick={[Function]}
  >
    Mark as Complete!
  </div>
</div>
`;

exports[`<ProjectList /> renders populated list 1`] = `
<ContextProvider
  value={
    Object {
      "error": null,
      "loading": false,
      "projects": Array [
        Object {
          "address": Array [
            "217 East Wall Road",
            "Dublin 19",
          ],
          "completed": false,
          "id": 9,
          "name": "Living Room Update",
          "startDate": "2019-04-24T00:00:00.000Z",
        },
        Object {
          "address": Array [
            "198 Fishamble Street",
            "Dublin 10",
          ],
          "completed": true,
          "id": 6,
          "name": "Window Replacements",
          "startDate": "2019-04-13T00:00:00.000Z",
        },
        Object {
          "address": Array [
            "99 Mountjoy Square",
            "Dublin 1",
          ],
          "completed": false,
          "id": 7,
          "name": "3 Bedroom Refurb",
          "startDate": "2019-04-13T00:00:00.000Z",
        },
      ],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ConnectFunction />
</ContextProvider>
`;

exports[`<ProjectList /> renders the loading spinner 1`] = `
<div
  className="spinner"
/>
`;
